ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,1409 @@
+            ; movzx eax, WORD [rsp + self.offset_trampoline as i32]
+            ; mov [rsp + self.offset_callee as i32], eax
+          ),
+          I(W) => dynasm!(self.assmblr
+            ; .arch x64
+            ; movsx eax, WORD [rsp + self.offset_trampoline as i32]
+            ; mov [rsp + self.offset_callee as i32], eax
+          ),
+          U(DW) | I(DW) => dynasm!(self.assmblr
+            ; .arch x64
+            ; mov eax, [rsp + self.offset_trampoline as i32]
+            ; mov [rsp + self.offset_callee as i32], eax
+          ),
+          U(QW) | I(QW) => dynasm!(self.assmblr
+            ; .arch x64
+            ; mov rax, [rsp + self.offset_trampoline as i32]
+            ; mov [rsp + self.offset_callee as i32], rax
+          ),
+        }
+        // Section 3.2.3 of the SysV AMD64 ABI:
+        // > The size of each argument gets rounded up to eightbytes. [...] Therefore the stack will always be eightbyte aligned.
+        self.offset_trampoline += 8;
+        self.offset_callee += 8;
+
+        debug_assert!(
+          self.allocated_stack == 0
+            || self.offset_callee <= self.allocated_stack
+        );
+      }
+    }
+  }
+
+  fn zero_first_arg(&mut self) {
+    dynasm!(self.assmblr
+      ; .arch x64
+      ; xor edi, edi
+    );
+  }
+
+  fn cast_return_value(&mut self, rv: NativeType) {
+    // V8 only supports 32bit integers. We support 8 and 16 bit integers casting them to 32bits.
+    // In SysV-AMD64 the convention dictates that the unused bits of the return value contain garbage, so we
+    // need to zero/sign extend the return value explicitly
+    match rv {
+      NativeType::U8 => dynasm!(self.assmblr; .arch x64; movzx eax, al),
+      NativeType::I8 => dynasm!(self.assmblr; .arch x64; movsx eax, al),
+      NativeType::U16 => dynasm!(self.assmblr; .arch x64; movzx eax, ax),
+      NativeType::I16 => dynasm!(self.assmblr; .arch x64; movsx eax, ax),
+      _ => (),
+    }
+  }
+
+  fn allocate_stack(&mut self, params: &[NativeType]) {
+    let mut int_params = 0u32;
+    let mut sse_params = 0u32;
+    for param in params {
+      match param {
+        NativeType::F32 | NativeType::F64 => sse_params += 1,
+        _ => int_params += 1,
+      }
+    }
+    let mut stack_size = (int_params.saturating_sub(Self::INTEGER_REGISTERS)
+      + sse_params.saturating_sub(Self::FLOAT_REGISTERS))
+      * 8;
+
+    // Align new stack frame (accounting for the 8 byte of the trampoline caller's return address)
+    // Section 3.2.2 of the SysV AMD64 ABI:
+    // > The end of the input argument area shall be aligned on a 16 (32 or 64, if
+    // > __m256 or __m512 is passed on stack) byte boundary. In other words, the value
+    // > (%rsp + 8) is always a multiple of 16 (32 or 64) when control is transferred to
+    // > the function entry point. The stack pointer, %rsp, always points to the end of the
+    // > latest allocated stack frame.
+    stack_size += (16 - (stack_size + 8) % 16) % 16;
```

Comment:
> But this same note, because the expression comes up as often as it does, it might be best as a utility function.

agreed

> I believe that the inner modulo here is unnecessary?

How exactly do you suggest to change the expression?

Reply:
> > But this same note, because the expression comes up as often as it does, it might be best as a utility function.
> 
> agreed
> 
> > I believe that the inner modulo here is unnecessary?
> 
> How exactly do you suggest to change the expression?

Afaik, it should be equivalent to
```rust
s += (16 - (s + 8)) % 16
```

Reply:
Still not seeing how this expression is equivalent, though. here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=12daa48d435d1708439591ad24df5844

Reply:
> Still not seeing how this expression is equivalent, though. here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=12daa48d435d1708439591ad24df5844

It doesn't work for signed integers because `%` is the remainder, and wrapping works differently. `elucid_rem` might work the same.

Use `55_u32` and it'll work, but it requires `u32::wrapping_sub`; this will reveal how they're equivalent, but I wouldn't suggest using the expression now because it requires an explicit wrapping subtraction, and LLVM optimises them into the same.